lib/function-promoter.js

```diff
@@ -0,0 +1,67 @@
+const validateArgumentTransformers = require('./argument-transformer-validator');
+const RiffError = require('./riff-error');
+const MappingTransform = require('./mapping-transform');
+const logger = require('util').debuglog('riff');
+
+const withTransformers = (promotedFunction, userFunction) => {
+    const transformers = userFunction['$argumentTransformers'];
+    if (typeof transformers === 'undefined') {
+        return promotedFunction;
+    }
+    validateArgumentTransformers(transformers);
+    const transformerCount = transformers.length;
+    if (transformerCount !== 1) {
+        throw new RiffError(
+            'error-argument-transformer',
+            `Request-reply function must declare exactly 1 argument transformer. Found ${transformerCount}`
+        );
+    }
+    promotedFunction['$argumentTransformers'] = transformers;
+    return promotedFunction;
+};
+
+const withHooks = (promotedFunction, userFunction) => {
+    const initHook = userFunction['$init'];
+    validateHook(initHook, 'init');
+    if (typeof initHook === 'function') {
+        promotedFunction['$init'] = userFunction['$init'];
+    }
+
+    const destroyHook = userFunction['$destroy'];
+    validateHook(destroyHook, 'destroy');
+    if (typeof destroyHook === 'function') {
+        promotedFunction['$destroy'] = userFunction['$destroy'];
+    }
+    return promotedFunction;
+};
+
+const validateHook = (hook, hookName) => {
+    const hookType = typeof hook;
+    if (hookType !== 'function' && hookType !== 'undefined') {
+        throw new RiffError('error-promoting-function', `Request-reply function ${hookName} hook must be a function. Found: ${hookType}`);
+    }
+};
+
+module.exports = (userFunction) => {
+    const interactionModel = userFunction['$interactionModel'] || 'request-reply';
```

Comment:
future: I'd expect to see a promoter for each interaction model with an error for an unknown model. This could lead to a extensible model where other interaction models are installed into the invoker as a dependency (or contributed by a buildpack).

Reply:
Renamed `function-promoter.js` to the hopefully more explicit `request-reply-promoter.js`.